apps/yggdrasil/lib/yggdrasil/user.ex

```diff
@@ -29,14 +36,133 @@ defmodule Yggdrasil.User do
   end
 
   @doc """
-  adds password at end of chain protects the hashpwsalt from seeing a nil value
-  in the case of missing password field.
+  Creates the user and assigns the default roles of "player"
+  """
+  def create_with_default_role(attributes) do
+    changeset = User.create_changeset(%User{}, attributes)
+
+    with {:ok, user} <- Repo.insert(changeset),
+          :ok        <- assign_role(user, "player"),
+     do: {:ok, user}
+  end
+
+  def assign_role(user, role_name) do
+    role = Repo.get_by!(Role, name: role_name)
+
+    user_role = UserRole.changeset(%UserRole{},
+                                   %{user_id: user.id, role_id: role.id})
+    case Repo.insert(user_role) do
+      {:ok, _role} -> :ok
+      error        -> error
+    end
+  end
+
+  def load_permissions(user) do
+    query = from ur in UserRole,
+            where: ur.user_id ==  ^user.id,
+            preload: [role: [:role_permissions]]
+
+    permissions = query
+    |> Repo.all()
+    |> Enum.flat_map(fn ur -> ur.role.role_permissions end)
+    |> Enum.group_by(fn rp -> rp.resource end)
+    |> Enum.map(fn {r, rps} -> {String.to_atom(r), Enum.map(rps, fn rp -> String.to_atom(rp.permission) end)} end)
+    |> Enum.into(%{})
+
+    %{ user | permissions: permissions}
+  end
+
+  @doc ~S"""
+  Checks the list of resource/permissions pair given against what the user
+  has been granted and returns true if and only if all permissions supplied
+  have been granted for the user.
+
+  ## roles
+  A user can have one or more roles with each role having a set of resources
+  and a set of resources and those resources having a set of permissions. Roles
+  are not checked directly and only serve as a grouping mechanism for resources.
+
+  Since the user can have multiple roles, the permission set checked for a given
+  resource is the unique set across all the roles for that resource.
+
+  So if the user has `Role1` and `Role2` defined as follows:
+
+  * Role1
+    * foo
+      * :read
+      * :write
+  * Role2
+    * foo
+      * :read
+      * :all
+
+  The set of permissions checked for foo will be `[:read, :write, :all]`
+
+  ## Examples
+      # checking a single resource
+      User.is_granted? user, foo: [:read]
+
+      # checking multiple resources is allowed as well
+      User.is_granted? user, foo: [:read], bar: [:read, :write]
   """
-  def hash_password(changeset = %{:valid? => false}) do
+  def is_granted?(_user, []) do
+    {:error, "res_perm list was empty"}
+  end
+
+  def is_granted?(user, res_perms) do
```

Comment:
Ok now I see why you keep talking about errors.

I was sticking to a true/false semantic where true is granted and false was not. This will break about every test I have but I can see where you are going. Errors in my implementation would literally not be something a user should ever see or we would want the user to see it's not `:no_access` it's the code is burning down something is horrible wrong and it's dying. :smile: hence why I started with exceptions. However, with this model I see what you are suggesting just was on a totally different page.

The intended use then would be something like 

```Elixir
case User.is_granted(user, [foo: [:bar]]) do
  :ok -> # something meaningful and say it's not in the plug display something
  {:error, _reason} -> # I don't have access do something else so in this case it's can I see all characters or not.
end

# however I could do this to show the user or someone why
case User.is_granted?(user, [foo: [:bar]]) do
  :ok -> # nothing to see here
  {:error, reasons} -> # render it some how.
end
```
That said there are still edge cases here where we want different errors I think, so something like

```elixir
:ok #all is good
{:error, :no_access, reasons} # a map of resource/reasons maybe like you have to render
{:error, :invalid_params, details} # invalid something anyway but the edger cases I test for which may not be something the user needs to see...
```